### PR TITLE
[FLINK-37281] Improve sink tests

### PR DESCRIPTION
**Pull request by AHeise**
Changes (test-only)

Refactor KafkaSinkITCase

KafkaSinkITCase's assertions are intermixed with operations making it hard to reason about the state and also introducing some flakiness in edge cases.

Improve extensibility in KafkaWriterTestBase

To test recovery in future PRs, it's important to decompose the #createWriter methods into common cases and advanced cases that may require some additional setup.

### CODE REVIEW

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaSinkITCase.java [hunk 1]
```diff
@@ -606,29 +682,21 @@ public void initializeState(FunctionInitializationContext context) throws Except
     private static final class InfiniteIntegerSource
             implements SourceFunction<Long>, CheckpointListener, CheckpointedFunction {
```
Comment (fapaul) [inline flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaSinkITCase.java:702]: Nit: Does this source still need to be a `CheckpointedFunction`? Most of the logic moved to the mapper.

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterTestBase.java [hunk 2]
```diff
@@ -227,6 +246,13 @@ public ProducerRecord<byte[], byte[]> serialize(
             }
             return new ProducerRecord<>(topic, ByteBuffer.allocate(4).putInt(element).array());
         }
+
+        @Override
+        public Optional<KafkaDatasetFacet> getKafkaDatasetFacet() {
+            return Optional.of(
+                    new DefaultKafkaDatasetFacet(
+                            DefaultKafkaDatasetIdentifier.ofTopics(Arrays.asList(topic))));
+        }
```
Comment (fapaul) [inline flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterTestBase.java:255]: Why was this necessary to add?

**Review by fapaul — APPROVED** [review]
LGTM % inline comments

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterTestBase.java (hunk 2 above)
Comment (AHeise) [reply flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterTestBase.java:255]: Not necessary for this PR, will revert.
